src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -161,6 +165,25 @@ private Course tryGetCourse(@NotNull Project project, @NotNull URL courseUrl) {
     }
   }
 
+  private void startAutoInstalls(@NotNull Project project,
+                                 @NotNull Course course,
+                                 boolean restartWhenFinished) {
+    ComponentInstaller.Factory factory
+        = new ComponentInstallerImpl.FactoryImpl<>(new SimpleAsyncTaskManager());
+    ComponentInstaller installer = factory.getInstallerFor(course);
+
+    Runnable callback = null;
+    if (restartWhenFinished) {
+      callback = () -> ApplicationManager.getApplication().invokeLater(() -> {
+        if (dialogs.showRestartDialog(project)) {
+          ideRestarter.restart();
+        }
+      });
```

Comment:
Nit: This is a bit scary with so many lambdas 😅 . Would it be an overkill to have a class such as
```java
public class PostponedRunnable implements Runnable {
  private final Runnable task;
  private final Executor invoker;

  /* For testing purposes */
  PostponedRunnable(Runnable task, Executor invoker) {
    this.task = task;
    this.invoker = invoker;
  }

  public PostponedRunnable(Runnable task) {
    this(task, ApplicationManager.getApplication()::invokeLater);
  }

  @Override
  public void run() {
    invoker.execute(task);
  }
}
```
and then write `callback = new PostponedRunnable(() -> { ...`

Reply:
I agree with @OlliKiljunen here, please, think of how this can be simplified (I don't have any ideas/proposal how) :confounded:

Reply:
I'm getting flashbacks to `FunctionUtil.lessThanOrEqualTo` :smile: But I admit, this one gets confusing to read with the multiple lambdas. I'll most likely go with that `PostponedRunnable` proposal :+1: